[CS-2772] Android backup & restore to Google Cloud

lukemelia opened the pull request:
### Description

This PR updates libraries and code around backup/restore, including drawing on a related PR from Rainbow with updates to react-native-cloud-fs.

I added a tappable option item to look for backups to the Import Account sheet on Android, since  connecting to Google Drive involves interaction with the device's Google account and the Drive OAuth flow. After tapping and connecting, if backups are found, the user enters the restore flow. If no backups are found, the item updates to reflect that state.

I updated most, but not all of the files involved to Typescript. 

### Checklist

- [x] All UI changes have been tested on a small device

### Screenshots

Add an existing account:

<img width="300" src="https://user-images.githubusercontent.com/353/146051582-ecb5253e-518f-4f3b-8b4c-22a7adb14971.png">

Restore a backup:

<img width="300" src="https://user-images.githubusercontent.com/353/146051641-583fc894-62d4-419a-a0b9-0e62422ca015.png">

Restoring:

<img width="300" src="https://user-images.githubusercontent.com/353/146051751-deb97b5d-d87f-4782-ac0d-5bcef1b87032.png">

Imported account, not backed up:

<img width="300" src="https://user-images.githubusercontent.com/353/146052167-1aff5aa9-9d46-415d-bd6f-b5d3ffd542bd.png">

New account, not backed up:

<img width="300" src="https://user-images.githubusercontent.com/353/146052959-05414de5-75a2-4297-9300-20b55b1957c3.png">

Choose a backup password:

<img width="300" src="https://user-images.githubusercontent.com/353/146052266-fa988c13-ad00-44d5-91db-e1a4a1afdca9.png">

<img width="300" src="https://user-images.githubusercontent.com/353/146052361-1810aba2-dffe-4c92-86d0-b74fbc3522ea.png">

Backing up...

<img width="300" src="https://user-images.githubusercontent.com/353/146052456-2c27ca8f-f84a-4432-af7a-daf382add87d.png">

Account is backed up:

<img width="300" src="https://user-images.githubusercontent.com/353/146051172-b9cd7420-b945-4782-b240-ae927070b6de.png">

Manage Google Drive backups: 

<img width="300" src="https://user-images.githubusercontent.com/353/146051351-629f0957-6243-4fb4-9b37-58441fb5dafd.png">

brunobar79 committed:
Cloud Backups Upgrade (#2071)

* support multiple iCloud backups

* bump react-native-cloud-fs

* better error handling + sync

* sync

* fix backup selection

* clean up

* update podfile lock

* backup ts

* remove old logs

* remove sync check

lukemelia committed:
Add option to try restoring from Google Cloud to Import Sheet

lukemelia committed:
Move cloudPlatform const to Device util

lukemelia committed:
Convert some Backup/Restore-related code to TypeScript

lukemelia committed:
Show "No backups found" message when No backups found in Google Drive.

lukemelia committed:
Hide password fields while backing up

lukemelia committed:
Upgrade react-native-mail, which now includes types

lukemelia committed:
Address lint warning

danibonilha commented on cardstack/src/types/react-native-dotenv.d.ts:
```diff
@@ -8,6 +8,7 @@ declare module 'react-native-dotenv' {
   export const CRYPTOCOMPARE_API_KEY: string;
   export const MERCHANT_ID: string;
   export const MERCHANT_ID_TEST: string;
+  export const RAINBOW_MASTER_KEY: string;
```
Maybe we change the name of this env key? So we won't reference rainbow, and I assume the one we have on our local .env it's a diff key from rb's one, not sure how to check though, if we were to rename this we can't forget to sync the secrets before publishing the new name

lukemelia replied:
Renamed in https://github.com/cardstack/cardwallet/pull/574/commits/e7140e8b18fef7aeaf6779d00873d894b4765463

danibonilha commented on src/components/backup/RestoreCloudStep.js:
```diff
@@ -67,16 +87,54 @@ export default function RestoreCloudStep({ userData }) {
+            logger.log('initializing wallet');
+          } else {
+            replace(Routes.SWIPE_LAYOUT);
+          }
```
Maybe this piece could be in it's own function, I find a bit hard to understand what it does just by looking at it, maybe inside the useWalletBackups or something

lukemelia replied:
Do you mean the post-backup navigation, or the entire onSubmit handling?

danibonilha replied:
I mean from line 109 to 130, I'm guessing is updateWallestBackupState ?

lukemelia replied:
I think I'll leave this as is for now. Considering how close it is to your vacation, I'd rather have you do another review pass and get this merged than delay further.

danibonilha commented on src/components/backup/RestoreSheetFirstStep.tsx:
```diff
@@ -42,6 +50,31 @@ export default function RestoreSheetFirstStep({
+          iconProps={{
+            name: 'download-cloud',
+            size: 25,
+          }}
```
we could extract this object outside the component function and use on both Option items

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/828af3c6934c6762463d506d8140564cbe556793

danibonilha commented on src/components/backup/RestoreSheetFirstStep.tsx:
```diff
@@ -42,6 +50,31 @@ export default function RestoreSheetFirstStep({
+            size: 25,
+          }}
+          marginTop={4}
+          onPress={() => {}}
```
now that we have a case where OptionItem is rendered without an action, maybe we could make `onPress` optional on `OptionalItem`. and have it disabled, if there's no `onPress`

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/fdb917a9db77a26d50c1af42fc6f62657d70c1b4

danibonilha commented on src/components/layout/Flex.ts:
```diff
@@ -25,17 +25,34 @@ const flexPropTypes = {
+  shrink?: number | null;
+  wrap?: boolean | null;
+}
+
```
Nice ts migration, but I wouldn't spend too much time on that, we are moving away from styled-components, every UI refactor possible we are removing the styled-components ones to have less styles deps, as today we have styled-components, restyle, vanilla react native styles.

lukemelia replied:
Cool. Can you point me at an ideal example of what we want to move toward?

danibonilha replied:
Of course, let's say we have a Row component  with styled components, we could replace that with our Container component:

```tsx
// StyledComponent
const Row = styled(Flex).attrs({ direction: 'row' });

// RestyleComponent
<Container flexDirection='row' flex={1} />

// RN vanilla
const styles = StyleSheet.create({
    container: {
       flexDirection: 'row',
       flex: 1
    }
});

// inside render
<View style={styles} />
```


Or for example you have a local styled-component component in your file, 

```tsx
const Wrapper = styled(Centered).attrs({
  align: 'start',
})`
  ${({ type }) =>
    type === BiometryTypes.FaceID || type === BiometryTypes.Face
      ? `
        margin-bottom: 2;
        margin-left: 4;
      `
      : `
        margin-bottom: 8;
        margin-left: 0`;
     }
`;
```
We could replace this with 

```tsx
// inside render to use condition
  const styles = useMemo(() => {
    const isFaceID = type === BiometryTypes.FaceID || type === BiometryTypes.Face
    
    return {
      marginBottom: isFaceID ? 2 : 8,
      marginLeft: isFaceID ? 4 : 0,
    }
  }, [type]);

<CenteredContainer alignItems='flex-start' style={styles}> 
```

This is the simplest example, but we can talk live about it if you want to.

lukemelia replied:
Thanks. That is helpful!

danibonilha commented on src/components/settings-menu/SettingsSection.js:
```diff
@@ -66,7 +66,7 @@ const checkAllWallets = wallets => {
-    if (!wallets[key].type !== WalletTypes.readOnly) {
+    if (wallets[key].type !== WalletTypes.readOnly) {
       canBeBackedUp = true;
     }
```
👀  good catch

danibonilha commented on src/hooks/useManageCloudBackups.js:
```diff
@@ -0,0 +1,142 @@
+  }, [dispatch, navigate, wallets]);
+
+  return { manageCloudBackups };
+}
```
I know it's legacy code but, would it be hard to have this as `.ts` or split  into smaller functions, as it is really confusing to understand ?

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/24b31af17ceeecc096568ca98dffb8c7867c28c5

danibonilha commented on src/model/backup.ts:
```diff
@@ -137,54 +145,81 @@ export function findLatestBackUp(wallets: AllRainbowWallets): string | null {
     if (!data) {
       throw new Error('Invalid password');
     }
+    let dataToRestore = {
+      ...data.secrets,
+    };
```
would `const dataToRestore = data.secrets` have the same behavior you expect for this case ?

lukemelia replied:
We're potentially mutating this object below on line 187. I think that's why it is using the rest operator to clone the object here. (This came from the Rainbow PR I merged in.)

danibonilha commented on src/components/backup/BackupCloudStep.tsx:
```diff
@@ -142,90 +156,93 @@ export default function BackupCloudStep() {
+                      name: biometryIconName as IconName,
+                    }
+                  : undefined
+              }
```
we could use[ this same approach](https://github.com/cardstack/cardwallet/blob/develop/cardstack/src/components/PinnedHiddenSection/PinHideOptionsFooter.tsx#L60) here

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/81097365e91769ab9bf6a392ba674164feca3d02

danibonilha commented on src/components/backup/BackupCloudStep.tsx:
```diff
@@ -142,90 +156,93 @@ export default function BackupCloudStep() {
+                      name: 'success',
+                    }
+                  : undefined
+              }
```
ditto

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/81097365e91769ab9bf6a392ba674164feca3d02

danibonilha commented on src/components/backup/BackupCloudStep.tsx:
```diff
@@ -142,90 +156,93 @@ export default function BackupCloudStep() {
+                      name: 'success',
+                    }
+                  : null
+              }
```
ditto

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/81097365e91769ab9bf6a392ba674164feca3d02

danibonilha commented on src/components/backup/BackupConfirmPasswordStep.tsx:
```diff
@@ -167,9 +172,26 @@ export default function BackupConfirmPasswordStep() {
+                    name: biometryIconName as IconName,
+                  }
+                : undefined
+            }
```
dito

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/6296f1d2b2bea908dfd1594102b3ca7b7546628f

danibonilha reviewed (commented):
Most of the comments are just improvements, I just need to test now behavior giving approval, but I'm letting my review so you can address what makes sense to you

danibonilha commented on src/screens/BackupSheet.js:
```diff
@@ -13,6 +13,8 @@ import {
 import { SlackSheet } from '../components/sheet';
 import { Container } from '@cardstack/components';
```
we need to remove this one

lukemelia replied:
https://github.com/cardstack/cardwallet/pull/574/commits/0b4b111e213bf610c064cb53568732a402213d0b

danibonilha commented on src/components/backup/BackupCloudStep.tsx:
```diff
@@ -142,90 +156,93 @@ export default function BackupCloudStep() {
+      </BackupSheetKeyboardLayout>
+    )
   );
 }
```
It seems some layout change broke the width of the inputs on this screen you can see the how it is on develop, on the second screenshoot [here](https://github.com/cardstack/cardwallet/pull/577), but just by looking at the code, no idea why it broke, may styled components update ?

danibonilha replied:
Which doesn't make much sense, because `RestoreCloudStep` uses the same component, and the input is fine, the only diff is that one is .js and the other .tsx

lukemelia replied:
<img width="300" src="https://user-images.githubusercontent.com/353/146263582-ab960aa5-17fd-4ad3-920f-b2c75a9d7b39.png">

https://github.com/cardstack/cardwallet/pull/574/commits/562fe79853c3358f5d6c78ad3e7fc88e03c5af52

danibonilha commented on src/screens/RestoreSheet.js:
```diff
@@ -15,17 +19,34 @@ import Routes from '@rainbow-me/routes';
+        // we didn't yet fetch the user data from the cloud
+        const data = await fetchUserDataFromCloud();
+        setParams({ userData: data });
+      }
```
Is there a diff place we can fetch this? My  concern is that we are awaiting for this request, and the user don't get any feedback, so I tap on `Have Cardwallet backup` and it takes a while for something to happen.
But if this is the only place to do this, we need some kind of loading, and also move the:
```
    // Animate transforming into backup sheet
    layoutEasingAnimation();
```
right before the last `setParams({ step: WalletBackupStepTypes.cloud });`, 
because `layoutEasingAnimation` triggers an animation on the next layout update, and we want this animation to happen while "navigating" to the next step, not when userData is updated, which happens before,  I'm even wondering if this `userData` needs to be set on the nav state, it seems to me, this could be a useRef, but this last piece we can check later.

lukemelia replied:
I added a loading state. I don't think there is a better spot considering that it can prompt interactive account/oauth UI.

https://github.com/cardstack/cardwallet/pull/574/commits/28b60f2e119d10d87d92257c559a185f646ca908

https://user-images.githubusercontent.com/353/146266068-a2627cb3-4e8f-4e19-af2b-971efa0e4a63.mov

danibonilha replied:
The loading state looks good ! 👍

lukemelia committed:
Rename RAINBOW_MASTER_KEY env var to CARDWALLET_MASTER_KEY

- I've synced a new secret with the correct name but have not yet removed the old one. We can do that once we merge and release.

lukemelia committed:
Extract duplicate icon props to a const

lukemelia committed:
make OptionItem's onPress optional, and render OptionItem as disabled if onPress is not supplied

danibonilha reviewed (changes requested):
Tested on Android, it's working good, but we broke the restore backup flow on iOS 😔 , keep getting
```
error getting userData [Error: item not found '/private/var/mobile/Library/Mobile Documents/iCloud~com~cardstack~cardpay/cardstack.com/wallet-backups/UserData.json']
```
And even inside the app, if I try to restore any of my previous backups while managing them, it does not restore successfully.

lukemelia committed:
Convert useManageCloudBackups to Typescript and reorganize

lukemelia committed:
Fix botched rebase

lukemelia committed:
Fix backup choice display name generation

lukemelia committed:
Extract iconProps definitions for Backup sheet options

lukemelia committed:
Simplify icon props in another part of backup UI

lukemelia committed:
Fix backup password fields layout

lukemelia committed:
Prettier

lukemelia committed:
Add loading state for looking for GDrive backups

lukemelia committed:
Fix how icloud backups are looked up

lukemelia commented:
@danibonilha iOS functionality should be fixed by https://github.com/cardstack/cardwallet/pull/574/commits/4566ae2f5471bb332ddee91ec11692ee866dbc97. Seems to be working for me.

danibonilha commented on package.json:
```diff
@@ -203,7 +203,7 @@
     "react-native-linear-gradient": "^2.5.6",
     "react-native-localize": "^2.1.5",
-    "react-native-mail": "^4.1.0",
+    "react-native-mail": "^6.1.1",
```
Native dependencies were updated, so we need to run `yarn install-pods` to update the iOS packages

danibonilha reviewed (approved):
Don't forget to push the `Podfile.lock` after `yarn install-pods` and it should be good to go!
Tested on Android and iOS it's looking good! Props on this one 👏

lukemelia committed:
Commit Podfile.lock to reflect updated pods
